src/txmempool.cpp

```diff
@@ -889,6 +982,140 @@ void CTxMemPool::removeForBlock(const std::vector<CTransactionRef> &vtx,
     std::list<CTransactionRef> &conflicts,
     bool fCurrentEstimate,
     std::vector<CTxChange> *txChanges)
+{
+    WRITELOCK(cs_txmempool);
+
+    // Process the block for transasction removal and ancestor state updates.
+    // As a first step remove all the txns that are in the block from the mempool by
+    // first updating the children for removal of the parent, and also saving the entries
+    // for policy estimate updates.  Then in the final loop simply remove all the txns.
+    //
+    // While we're updating the child transactions prior to removal we can also gather the
+    // mapTxnChaiTips.
+    mapEntryHistory mapTxnChainTips;
+    std::vector<CTxMemPoolEntry> entries;
+    setEntries setAncestorsFromBlock;
+    {
+        setEntries setTxnsInBlock;
+        for (const auto &tx : vtx)
+        {
+            uint256 hash = tx->GetHash();
+            txiter it = mapTx.find(hash);
+            if (it == mapTx.end())
+            {
+                continue;
+            }
+
+            // Get all ancestors from related to txns in the block. Stop looking if we've already looked up
+            // this set of ancestors before; we don't want to be traversing the same part of the ancestor
+            // tree more than once.
+            setEntries setAncestors;
+            uint64_t nNoLimit = std::numeric_limits<uint64_t>::max();
+            std::string dummy;
+            _CalculateMemPoolAncestors(
+                *it, setAncestors, nNoLimit, nNoLimit, nNoLimit, nNoLimit, dummy, &setAncestorsFromBlock, false);
+            setAncestorsFromBlock.insert(setAncestors.begin(), setAncestors.end());
+
+            setTxnsInBlock.insert(it);
+            entries.push_back(*it); // TODO: seems inefficient to do a full copy here?
```

Comment:
do you think we need to do something for the above TODO? or is it something we could do on a separate PR?

Reply:
This could be for another PR, it's not something I introduced here...been in the code a long time and isn't really all that much of a perf hit but It would be good to fix.

Reply:
We must copy the CTxMemPoolEntry structure here because it will be removed from the mempool in the next loop.  And "entries" is used later down.  But is only used in one place in the following code:

// After the txs in the new block have been removed from the mempool, update policy estimates
minerPolicyEstimator->processBlock(nBlockHeight, entries, fCurrentEstimate);

It does not seem to me that exactly when we update the mining policy estimator matters, because this operation is locked so seems atomic from the perspective of other mempool access code.  So we could move it up here, and it could be modified to take a set (setTxnsInBlock) rather than a vector, completely eliminating the need for "entries".

I am guessing that in practice this may have a big effect (now that we've fixed a lot of other stuff) on performance for large blocks because in my experience with optimization, I often see the code spending a lot of time copying transactions, specifically the scripts within a transaction (and BTW the "prevector" object is also an attempt to optimize this same area).  

Of course root cause of this qualitative assessment may be either because copying tx are slow OR because copying them happens a lot so its just a guess that optimizing this will be worthwhile.

Reply:
Seems like @ptschip is handling this in #2042. I'm pretty sure that he's leveraging `setTxnsInBlock` even thou I'm not sure that is doing the exact thing you are suggesting.

Reply:
handled in #2042